parsing.go

```diff
@@ -283,21 +314,164 @@ func parseReq(txt string) (*Req, error) {
 		return nil, fmt.Errorf("Requirement body must not be empty: %s", r.ID)
 	}
 
+	// PARENTS must be punctuation/space separated list of parseable req-ids.
+	err = parseParents(r)
+	if err != nil {
+		return nil, err
+	}
+
+	return r, nil
+}
+
+// parseReqTable reads a table of requirements one row at a time and parses the content into Req structures which are
+// then returned in a slice.
+//
+// Tables have the following format:
+// | ID | Title | Body | Attribute1 | Attribute2 |
+// | --- | --- | --- | --- | --- |
+// | ReqID | <text> | <text> | <text> | <text> |
+//
+// The first column must be "ID" and each row must contain a valid ReqID. Other columns are optional.
+//
+// @llr REQ-TRAQ-SWL-25
+func parseReqTable(txt string, reqs []*Req) ([]*Req, error) {
+
+	var attributes []string
+
+	// Split the table into rows and loop through
+	for rowN, rowS := range strings.Split(txt, "\n") {
+
+		// The first row contains the attribute names for each column, the first column must be "ID"
+		if rowN == 0 {
+			if reTableHeader.MatchString(rowS) {
+				attributes = splitTableLine(rowS)
+				for i, a := range attributes {
+					attributes[i] = strings.ToUpper(a)
+				}
+			} else {
+				return reqs, fmt.Errorf("first column must be \"ID\"")
+			}
+		} else {
+			if reTableDelimiter.MatchString(rowS) {
+				// Ignore the delimiter row
+				continue
+			}
+
+			values := splitTableLine(rowS)
+
+			if len(values) == 0 {
+				// End of table
+				break
+			}
+
+			if len(values) < len(attributes) {
+				return reqs, fmt.Errorf("too few cells on row %d of requirement table", rowN+1)
+			}
+
+			r := &Req{Attributes: map[string]string{}}
+
+			// For each attribute in the first row, read in the associated value on this row
+			for i, k := range attributes {
+				if k == "ID" {
+					ID, IDNumber, err := extractIDAndNum(values[i])
+					if err != nil {
+						return reqs, err
+					}
+					r.ID = ID
+					r.IDNumber = IDNumber
+				} else if k == "TITLE" {
+					r.Title = values[i]
+				} else if k == "BODY" {
+					r.Body = values[i]
+				} else {
+					if k == "PARENT" {
+						// make our lives easier, accept both, output only PARENTS
+						k = "PARENTS"
+					}
+
+					if values[i] != "" {
+						r.Attributes[k] = values[i]
+					}
+				}
+			}
+
+			err := parseParents(r)
+			if err != nil {
+				return reqs, err
+			}
+
+			reqs = append(reqs, r)
+		}
+	}
+
+	return reqs, nil
+}
+
+// splitTableLine splits a pipe table row in cells. Does not count for
+// escaped `|` characters or other cases when `|` should not be considered
+// a cell separator. Removes the first and last parts if they are empty.
+func splitTableLine(line string) []string {
+	// The `|` at the beginning of the line is ignored because it
```

Comment:
I don't think this is worth doing, but maybe worth discussing:
you have reTableHeader, which with a little work would be able to
also do this whole function's job. But adding yet more regular expressions
might just be the proverbial straw :)

Reply:
yeah, I think you're right, but I couldn't find a way to extract all the cell information from a table row and remove any outer spaces. I always ended up having to loop through the results and call TrimSpace, like this: https://play.golang.org/p/eoqvKMUHido